service/src/test/java/bio/terra/workspace/common/utils/MdcHookTest.java

```diff
@@ -14,147 +19,228 @@ class MdcHookTest extends BaseUnitTest {
-    stairway.submit(flightId, MdcFooBarUndoTestFlight.class, flightMap);
-    Thread.sleep(3000);
-
     // CheckMDC passes on do & undo, otherwise the flight would end with FlightStatus.FATAL. The
     // ErrorStep causes the undo to happen and the flight to end as ERROR.
-    assertEquals(FlightStatus.ERROR, stairway.getFlightState(flightId).getFlightStatus());
+    var flightState = submitAndWait(MdcFooBarUndoTestFlight.class);
+    assertEquals(
+        FlightStatus.ERROR,
+        flightState.getFlightStatus(),
+        "Flight fails but all steps can be walked back");
+    var maybeFlightException = flightState.getException();
+    assertTrue(maybeFlightException.isPresent());
+    assertEquals(
+        ErrorStep.EXPECTED_EXCEPTION.getMessage(),
+        maybeFlightException.get().getMessage(),
+        "Flight fails due to ErrorStep");
+    assertEquals(
+        FOO_BAR, MDC.getCopyOfContextMap(), "Calling thread's context is preserved after flight");
   }
 
   @Test
-  void noInitialContextOk() throws Exception {
-    Stairway stairway = jobService.getStairway();
-
-    MDC.clear();
-    FlightMap flightMap = new FlightMap();
+  void emptyInputContextOk() throws Exception {
     flightMap.put(MdcHook.MDC_FLIGHT_MAP_KEY, mdcHook.getSerializedCurrentContext());
 
-    String flightId = stairway.createFlightId();
-    stairway.submit(flightId, MdcNoContextTestFlight.class, flightMap);
-    Thread.sleep(3000);
-
-    assertEquals(FlightStatus.SUCCESS, stairway.getFlightState(flightId).getFlightStatus());
+    assertEquals(
+        FlightStatus.SUCCESS,
+        submitAndWait(MdcNoInputContextTestFlight.class).getFlightStatus(),
+        "Flight succeeds when empty input context provided");
+    assertNull(MDC.getCopyOfContextMap(), "Calling thread's context is preserved after flight");
   }
 
   @Test
   void inputParametersNotSetOk() throws Exception {
-    Stairway stairway = jobService.getStairway();
-
-    String flightId = stairway.createFlightId();
-    // Don't set the MDC_FLIGHT_MAP_KEY on the input FlightMap.
-    stairway.submit(flightId, MdcNoContextTestFlight.class, new FlightMap());
-    Thread.sleep(3000);
-
-    assertEquals(FlightStatus.SUCCESS, stairway.getFlightState(flightId).getFlightStatus());
+    assertEquals(
+        FlightStatus.SUCCESS,
+        submitAndWait(MdcNoInputContextTestFlight.class).getFlightStatus(),
+        "Flight succeeds when no input context provided");
+    assertNull(MDC.getCopyOfContextMap(), "Calling thread's context is preserved after flight");
   }
 
   /** Test that the CheckMdc Step will fail the flight as expected by the rest of the tests. */
   @Test
   void testCheckMdc() throws Exception {
-    Stairway stairway = jobService.getStairway();
-
-    String flightId = stairway.createFlightId();
     // Don't set the MDC_FLIGHT_MAP_KEY on the input FlightMap, but expect "foo": "bar"
-    stairway.submit(flightId, MdcFooBarTestFlight.class, new FlightMap());
-    Thread.sleep(3000);
```

Comment:
👍 nice that is shifting to a less flaky submitAndWait rather than the arbitrary wait for 3 seconds.